VisualShader: Add a CoordinateSpaceHelper node and add proper `mat4 * vec4` operation to TransformVecMult node

tetrapod00 opened the pull request:
Partially implements https://github.com/godotengine/godot-proposals/issues/7666.
Resolves https://github.com/godotengine/godot-proposals/issues/8376.

**Problem summary:**
The current visual shader node [TransformVecMult](https://docs.godotengine.org/en/stable/classes/class_visualshadernodetransformvecmult.html) is trying to serve two purposes: a way to transform vectors between coordinate spaces, and also a general `vec4 * mat4` node. It's failing at both purposes:
- It has correct but obscure dropdown names like `A x B (3 x 3)`, making it hard to understand if you don't already know matrix math.
- It can't actually do general-purpose `vec4 * mat4` multiplication correctly, since it accepts a `vec3` and autofills the `w` component of a `vec4` with either `1.0` or `0.0`.
- It doesn't output a `vec4`, so you can't use the `w` component of the output. (Useful for some clip space operations).
- It's not beginner or artist friendly. Other engines have dedicated coordinate space transform nodes, but godot requires you to know at least some matrix math.

**Solution:**
Add a new, dedicated node to transform positions and directions between model, world, view, and clip space.
- Accepts and returns `vec3`. This node is *only* for transforming positions and directions. If you need to use the `w` component of the output, use the old node.
- Uses artist-friendly names of `Position` and `Direction` for the vector types.
- Automatically fills in the matrix needed internally. But, as an artist-friendly node, you don't need to know that.

Adjust the old `TransformVecMult` node. It now:
- Always outputs a `vec4`. If you only want a `vec3`, you can plug the `vec4` into a `vec3` input or swizzle the `vec4`.
- Keeps its four old modes, `A x B`, `B x A`, `A x B (3x3)`, `B x A (3x)`. These still accept a `vec3` input, but they now output the whole `vec4`. These modes function the same as before, except for outputting `w`.
- Adds two new modes (operations), `A x B (Vec4)` and `B x A (Vec4)`. These accept a `vec4` input and return a `vec4` output. If you want to use these modes to do the coordinate space transform, you'll need to fill in the correct `w` on the input.
- Backwards compatible, except for one edge case. If you are currently using the `vec3` output, plugging it into a `vec4` input, and *assuming* that the `w` component is `0.0`, your shader may change.

**New node:**
From spaces: `Model to...`, `World to...`, `View to...`, `Clip to...`
To spaces: `Model`, `World`, `View`, `Clip`
Vector types: `Position`, `Direction`
![godot windows editor dev x86_64_JBbHJ9swWh](https://github.com/user-attachments/assets/53df76ed-1304-4eba-aa6e-a49c9b8b6602)

**Visual summary of the two old and new node:**
![godot windows editor dev x86_64_L1OT8vxS5m](https://github.com/user-attachments/assets/0a9adbda-2b41-41ff-8c23-da8d9602de45)

**View to world snippet, old and new:**
![godot windows editor dev x86_64_i106PuUFHv](https://github.com/user-attachments/assets/c0bd7fbc-b5c7-4f59-821b-04f1ad9fe496)


**TransformVecMult compatibility**
<details>
<summary>Images of current compatibility</summary>

4.3: Shader that uses all four current operater modes, and connects to both `vec3` and `vec4` slots:

![Godot_v4 3-stable_win64_0emenX7GeH](https://github.com/user-attachments/assets/1c7fa79f-350f-4243-9159-2883b89fafe9)
This PR: Same shader, opened with this PR. All connections are still present. The only edge case is that the `vec3` output is now connected to a `vec4` slot. If your shader assumed that the `w` component is `0.0`, it will break.

![godot windows editor dev x86_64_E5X1InPsZf](https://github.com/user-attachments/assets/fb825b61-2e58-4654-8431-c35129fb47f6)

</details>

If 100% compatibility is needed, there are ways to do so. We could:
1. Output both a `vec3` and `vec4` for all operator modes of `TransformVecMult`, with the only difference being the `w` field. IIRC Unity did this for some of its nodes. It makes the node more cluttered, though.
2. Output either a `vec3` or a `vec4`, dynamically switched with another enum dropdown. Again, more clutter.
3. For the old operators, output a `vec3`. For the new `vec4 * mat4` operators, output a `vec4`. Old operators remain exactly as they are, and the correct `vec4 * mat4` operator is now available if you need `w` or for other uses.
4. Alternately, we could use a different compatibility strategy. If I was designing the nodes today, without compatibility concerns, I would make two nodes: `CoordinateSpaceHelper` (as described in this PR) and a `TransformVector4Multiply` node, which *only* does `vec4 * mat4` or `mat4 * vec4` multiplications.  The `CoordinateSpaceHelper` does artist-friendly transforms, and the `TransformVector4Multiply` node matches text shader behavior (since you can't do `vec3 * mat4` or `mat4 * vec3` in a text shader).
So we could add those two nodes, and deprecate the current node. Users of the current `TransformVecMult` should switch to either `CoordinateSpaceHelper` for the common use case, or `TransformVector4Multiply` for more any other use case.

**Tasks:**
- [x] Decide on the new node's name. "CoordinateSpaceHelper" is intentionally not the final name.
- [x] Decide on names for new enums. The names for `TransformVecMult`'s operators should maybe be changed, too.
- [ ] Test old node for compatibility.
- [ ] Test more thoroughly that each coordinate space conversion works in vertex(), fragment(), and light(). 
- [x] Write docs.

tetrapod00 committed:
Initial

Working version

Fix copy paste error

Change TransformVecMult to output vec4, generate docs

Some more docs.

Suggestion: remove unnecessary default

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

Chaosus reviewed (dismissed):
LGTM, A minor compatibility breakage is acceptable in that case (I think)

tetrapod00 commented:
We probably should find a way to actually keep 100% compatibility. Having a single node that both does `mat4 * vec4` and `mat4 * vec4(vec3, 1.0)` is already something of a compromise from the ideal set of nodes. So if we're not going with compatibility option 4 listed in the OP, we should probably pick option 1, 2, or 3. I think 3 is my favorite.

akien-mga commented:
This seems to still be a WIP, but next time you update it, please make sure to amend the commit message. It's currently:
![image](https://github.com/user-attachments/assets/d9057532-bf0a-42ce-9170-2c6629cca86c)

"Initial" isn't a proper commit title :) And the messages from intermediate squashed commits aren't relevant to the final commit and should be removed from the description.

QbieShay commented:
I'm not opposed to the idea of deprecating the current node. We can still keep it available as a class, but not available anymore in the node list. This should make it impossible for people to add it to a new shader.

tetrapod00 commented:
> I'm not opposed to the idea of deprecating the current node. We can still keep it available as a class, but not available anymore in the node list. This should make it impossible for people to add it to a new shader.

That would be my preferred option. Just so you're aware, it does make the use case described in https://github.com/godotengine/godot-proposals/issues/8376 slightly more effort to setup. If you want to multiply INV_PROJECTION_MATRIX and a vector and use the `w` output, you now have to know the whole setup of `vec4 view = (INV_PROJECTION_MATRIX * vec4(clip_position, 1.0)`. In the new visual shader with option 4, that would look something like:
![godot windows editor x86_64_P47hyb6ebF](https://github.com/user-attachments/assets/1054fae2-2ae0-47f7-ad82-099a5b9d7417)

I do think this is okay. Visual shader should support the same, correct mathematical operators that text shaders do, and also make common use cases easier with high level nodes. Going with option 4 and deprecating the current node seems like the right long-term solution to me.

tetrapod00 commented:
> I do think for the linearized scene depth we should just offer a high level node
We already do! We also offer one for reconstructing world coordinates from the screen uv and depth, which I believe are the most common two use cases for using the `w` component.

tetrapod00 committed:
Visual Shader: Add new TransformVectorCoordinate and TransformVectorMultiply nodes, deprecate TransformVecMult node

tetrapod00 commented:
I implemented option 4 from the OP: add two new nodes, and deprecate the old node.

To reiterate: long term, the right set of nodes is one node which only does `vec4 * mat4` and `mat4 * vec4`, and one node which only does coordinate transformation for `vec3`.

Between the two nodes, you can do anything that you currently can. In most cases, you can replace existing uses of `TransformVecMult` with a `VectorCoordinateTransform` node and use fewer nodes. In some cases, if you're using custom transform matrices, you can use the new `TransformVectorMultiply` node, and you may need to add some extra nodes to set it up.

I marked the old node as deprecated in the class ref. I also added a deprecation message to the node itself. I did not remove the old node from the `Add Node` menu. I don't know if there is a precedent for deprecating VisualShader nodes, but IMO the node should be addable for at least one minor version, then removed from the `Add Node` list. Alternately, we can immediately deprecate the current node in 4.4.

All existing instances of the current node in existing visual shaders will continue to function with no changes.

Summary of new and old nodes:
![godot windows editor dev x86_64_BpKlsqbQAF](https://github.com/user-attachments/assets/85c2c4ca-b9e8-4e5a-92ea-0304b44dfeb4)

**Naming:**
- `TransformVectorMultiply` is named verbosely. Most other visual shader nodes spell out "Vector", and the `MultiplyAdd` node also spells out "Multiply". This also helps distinguish it from the old node with a similar name.
- `VectorCoordinateTransform` is what I decided on for the new node, but I'm open to suggestions. "Vector" because that is the type being operated on, "Coordinate" to specify it's about coordinate spaces, and "Transform" because "transform" is the verb in this case. 
    - I also considered "TransformVectorCoordinate", "VectorTransform", "TransformVector", "VectorCoordinateConversion". Feel free to suggest other names.
    - `TransformVectorCoordinate.VectorType` is named somewhat verbosely, `VECTOR_TYPE_POSITION` and `VECTOR_TYPE_DIRECTION`. Arguably this could be `VECTOR_POSITION` or `TYPE_POSITION`. These are not the user-visible names within the graph itself, and the user-visible names can be changed easily.

tetrapod00 commented:
As pointed out in https://github.com/godotengine/godot-proposals/issues/7666#issuecomment-2429682953, this PR currently doesn't handle clip space correctly. Handling clip space correctly will involve inputting and/or outputting a vec4 instead of a vec3, and the current design of two cleanly distinguished nodes, one that does `vec3<->vec3`, and one that does `mat4 * vec4` node, doesn't quite work. 

Because of that and because the current PR involves "deprecating" a visual shader node, I'm marking this as a draft again to reconsider the overall design. The code for the `VectorCoordinateTransform` node in this PR should still be mostly usable for `model`, `world`, and `view` space transformations. I'm not sure when I'll pick this up again. Anyone else can salvage this if they want to.
